### ISSUE TITLE
Enhancement: [no-unnecessary-condition] Allow granualar configuration - encourages non-exhaustive enum/union checking

### ISSUE THREAD

**Issue by phiresky** [typescript-eslint/typescript-eslint#7561]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition/

### Description

This rule is amazing, it catches a ton of huge issues in our codebase. 

But: the "both sides of the expression are literal values" rule seems to encourage a bad pattern where tagged unions and enum if-else chains are discouraged to be complete, instead encouraging open else-cases.

no-unnecessary-condition is really a collection of 10 different rules:
```ts
            alwaysTruthy: 'Unnecessary conditional, value is always truthy.',
            alwaysFalsy: 'Unnecessary conditional, value is always falsy.',
            alwaysTruthyFunc: 'This callback should return a conditional, but return is always truthy.',
            alwaysFalsyFunc: 'This callback should return a conditional, but return is always falsy.',
            neverNullish: 'Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined.',
            literalBooleanExpression: 'Unnecessary conditional, both sides of the expression are literal values.',
            ...
```

Since catching the tagged-union case above is probably pretty of hard, I propose each of these to be on-off configurable individually. I've gone through around 50 instances of this rule in our code base, and the `literalBooleanExpression` part was the only bad part.

### Fail

```typescript
(all the already caught ones)
```

### Pass

```typescript
enum TestStatus {
	pending = "pending",
	cancelled = "cancelled",
	success = "success",
}

function handle(status: TestStatus) {
	if (status === TestStatus.pending) {
		// handle pending
	} else if (status === TestStatus.cancelled) {
		// handle cancelled
	} else if (status === TestStatus.success) {
                // handle success
	} else {
		console.log(`Unhandled case ${String(expectNever(status))}`);
	}
}

/** ensure type error if argument is no longer unknown */
function expectNever(x: never): unknown {
	return x;
}
```

### Additional Info

The above code is (in my opinion) very clean. But this lint rule encourages removing the `never` else case and replacing `else if (status === TestStatus.success) {` with `else {`:

```diff
-	} else if (status === TestStatus.success) {
-                // handle success
	} else {
-		console.log(`Unhandled case ${String(expectNever(status))}`);
+                // handle success (and all other else cases)
+	}

```

That's fine right now, but it happens often that you add another option to to an enum or tagged union:

```diff
 enum TestStatus {
 	pending = "pending",
 	cancelled = "cancelled",
 	success = "success",
+ 	failed = "failed",
 }
```

With the code as above, this case would nicely be handled:

1. at runtime, the pre-coded else case would be hit
2. at compile-time, expectNever would trigger and require adding a new case to the if-else chain.

With the code as this lint rule currently encourages, this would happen:

1. at runtime, the new "failed" case would silently be handled like the "success" case
2. at compile time, nothing would indicate any issues.

**Comment by bradzacher** [comment]
This comes up a lot and unfortunately is just a downside of using `if`/`else` to exhaust a union type.

There's no way for us to know how you got to the state of only having one possible value for each side of the condition - all we know is that is the state you're in. Which is why we error.

Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/4682

You should use a switch instead of an `if` to exhaustively narrow a union - instead.

---

This isn't something we'll make configurable because there are many other cases where you actually do want this exact report.

For example imagine the following code:

```ts
function foo(arg: 1) {
  if (arg === 1) { // condition is unnecessary! 
    return;
  }
}
```

Obviously this is a vastly simplified example - but there are many, many cases where you would want the rule to error for this message.

The above issue contains more examples.

**Comment by phiresky** [comment]
@bradzacher I agree that there's multiple cases where the literal comparison lint issue *is* a true positive.

But going through our code base, there are ~500 instances of the `no-unnecessary-condition` hitting in general. Of that, only ~100 are the `literalBooleanExpression` case. Going through those, it seems like > 90% are the "false-positive" case as I described above.

So without configurability, I can either disable this rule and miss out of 410 instances of bad code. Or I can enable the rule and have ~90 cases where the code is now worse than before.

If this was configurable more granularily, I could find 400 of 410 instances of bad code while keeping the 90 instances of good code. I would only miss out on ~10 cases of bad code. **That's a reduction from 500 to 10 just by making this configurable**.

Would you be open to a PR that adds configurability to either just the one case or all different cases?
